# Deployment
## Input : `zone`, `approx_year_built`, `num_bedrooms`
## Output : `Sale Price`

## I. Train Model on Full Dataset

In [3]:
from sklearn.ensemble import BaggingRegressor
from sklearn.tree import DecisionTreeRegressor
import pandas as pd

df = pd.read_csv("housing_clean_imputed_data.csv")
df

sale_year  zone  cats_allowed  dogs_allowed  coop_condo  garage_exists  \
0       2016.00   7.0           1.0           1.0         1.0            0.0   
1       2016.00   2.0           0.0           0.0         2.0            0.0   
2       2016.08   6.0           1.0           1.0         2.0            0.0   
3       2016.12   3.0           1.0           1.0         1.0            0.0   
4       2016.28   9.0           0.0           0.0         1.0            0.0   
...         ...   ...           ...           ...         ...            ...   
2174    2016.00   3.0           0.0           0.0         1.0            0.0   
2175    2016.02   2.0           1.0           1.0         2.0            0.0   
2176    2016.00   6.0           0.0           0.0         1.0            0.0   
2177    2016.26   6.0           0.0           0.0         1.0            1.0   
2178    2016.24   2.0           0.0           0.0         1.0            0.0   

      num_bedrooms  num_floors_in_building  num_full_bathrooms  \
0              1.0                    2.00                 1.0   
1              2.0                    7.24                 2.0   
2              5.0                    2.00                 2.0   
3              1.0                    2.00                 1.0   
4              1.0                    6.00                 1.0   
...            ...                     ...                 ...   
2174           1.0                    2.00                 1.0   
2175           2.0                    7.34                 2.0   
2176           2.0                   10.60                 2.0   
2177           1.0                    6.00                 1.0   
2178           1.0                    4.18                 1.0   

      num_half_bathrooms  ...  dining_room_type  fuel_type  kitchen_type  \
0                    0.0  ...              3.00        3.0           4.0   
1                    0.0  ...              5.00        5.0           5.0   
2                    0.0  ...              3.00        3.0           5.0   
3                    0.0  ...              5.00        1.0           2.0   
4                    0.0  ...              5.00        1.0           2.0   
...                  ...  ...               ...        ...           ...   
2174                 0.0  ...              3.00        5.0           4.0   
2175                 0.0  ...              3.58        3.0           4.0   
2176                 0.0  ...              5.00        3.0           2.0   
2177                 0.0  ...              4.10        1.0           4.0   
2178                 0.0  ...              4.00        3.0           5.0   

      community_district_num  approx_year_built  sq_footage  walk_score  \
0                       26.0             1949.0      585.00        83.0   
1                       25.0             2007.0     1075.00        98.0   
2                       28.0             1951.0      980.00        77.0   
3                       25.0             1955.0      700.00        88.0   
4                       30.0             1940.0      659.68        99.0   
...                      ...                ...         ...         ...   
2174                    25.0             1950.0      764.84        67.0   
2175                    25.0             1987.0     1115.54        97.0   
2176                    24.0             1968.0     1134.90        88.0   
2177                    28.0             1939.0      100.00        98.0   
2178                    25.0             1950.0      643.92        97.0   

      maintenance_cost  parking_charges  sale_price  
0               529.00              0.0    170454.0  
1               720.52              0.0    565000.0  
2               800.34              0.0    386632.0  
3               658.00              0.0    158540.0  
4               783.00              0.0    155370.0  
...                ...              ...         ...  
2174            735.00              0.0    169000.0  
2175  

In [4]:
print("Splitting X and y ...")
# Combine training and validation sets
X = df.drop('sale_price', axis=1)
y = df['sale_price']

print("Training Best Model: Bagged Decision Trees")
# Best hyperparameters
best_params = {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 300}

# Initialize the Bagging Regressor with the best hyperparameters
best_model = BaggingRegressor(
    base_estimator=DecisionTreeRegressor(max_depth=best_params['max_depth'], 
                                         min_samples_split=best_params['min_samples_split'], 
                                         min_samples_leaf=best_params['min_samples_leaf']),
    n_estimators=best_params['n_estimators'], 
    random_state=42
)

print("Fitting model on entire dataset ...")
# Train the model on the entire dataset
best_model.fit(X, y)

print("Model training complete...")


c:\Users\lenovo\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\ensemble\_base.py:156: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


Model trained on the entire dataset.


In [19]:
X

sale_year  zone  cats_allowed  dogs_allowed  coop_condo  garage_exists  \
0       2016.00   7.0           1.0           1.0         1.0            0.0   
1       2016.00   2.0           0.0           0.0         2.0            0.0   
2       2016.08   6.0           1.0           1.0         2.0            0.0   
3       2016.12   3.0           1.0           1.0         1.0            0.0   
4       2016.28   9.0           0.0           0.0         1.0            0.0   
...         ...   ...           ...           ...         ...            ...   
2174    2016.00   3.0           0.0           0.0         1.0            0.0   
2175    2016.02   2.0           1.0           1.0         2.0            0.0   
2176    2016.00   6.0           0.0           0.0         1.0            0.0   
2177    2016.26   6.0           0.0           0.0         1.0            1.0   
2178    2016.24   2.0           0.0           0.0         1.0            0.0   

      num_bedrooms  num_floors_in_building  num_full_bathrooms  \
0              1.0                    2.00                 1.0   
1              2.0                    7.24                 2.0   
2              5.0                    2.00                 2.0   
3              1.0                    2.00                 1.0   
4              1.0                    6.00                 1.0   
...            ...                     ...                 ...   
2174           1.0                    2.00                 1.0   
2175           2.0                    7.34                 2.0   
2176           2.0                   10.60                 2.0   
2177           1.0                    6.00                 1.0   
2178           1.0                    4.18                 1.0   

      num_half_bathrooms  num_total_rooms  dining_room_type  fuel_type  \
0                    0.0              3.0              3.00        3.0   
1                    0.0              7.0              5.00        5.0   
2                    0.0              5.0              3.00        3.0   
3                    0.0              3.0              5.00        1.0   
4                    0.0              3.0              5.00        1.0   
...                  ...              ...               ...        ...   
2174                 0.0              3.0              3.00        5.0   
2175                 0.0              6.0              3.58        3.0   
2176                 0.0              4.0              5.00        3.0   
2177                 0.0              3.0              4.10        1.0   
2178                 0.0              3.0              4.00        3.0   

      kitchen_type  community_district_num  approx_year_built  sq_footage  \
0              4.0                    26.0             1949.0      585.00   
1              5.0                    25.0             2007.0     1075.00   
2              5.0                    28.0             1951.0      980.00   
3              2.0                    25.0             1955.0      700.00   
4              2.0                    30.0             1940.0      659.68   
...            ...                     ...                ...         ...   
2174           4.0                    25.0             1950.0      764.84   
2175           4.0                    25.0             1987.0     1115.54   
2176           2.0                    24.0             1968.0     1134.90   
2177           4.0                    28.0             1939.0      100.00   
2178           5.0                    25.0             1950.0      643.92   

      walk_score  maintenance_cost  parking_charges  
0           83.0            529.00              0.0  
1           98.0            720.52              0.0  
2           77.0            800.34              0.0  
3           88.0            658.00              0.0  
4           99.0            783.00              0.0  
...          ...               ...              ...  
2174        67.0            735.00              0.0  
2175  

In [15]:
import numpy as np


TypeError: predict_price() missing 1 required positional argument: 'property_type'

In [21]:
import numpy as np

# Example function to predict price based on input features
def predict_price(
    sale_year, zone, cats_allowed, dogs_allowed, coop_condo, garage_exists,
    num_bedrooms, num_floors_in_building, num_full_bathrooms, num_half_bathrooms,
    num_total_rooms, dining_room_type, fuel_type, kitchen_type, community_district_num,
    approx_year_built, sq_footage, walk_score, maintenance_cost, parking_charges
):
    # Combine the input features into a single array
    input_features = np.array([
        sale_year, zone, cats_allowed, dogs_allowed, coop_condo, garage_exists,
        num_bedrooms, num_floors_in_building, num_full_bathrooms, num_half_bathrooms,
        num_total_rooms, dining_room_type, fuel_type, kitchen_type, community_district_num,
        approx_year_built, sq_footage, walk_score, maintenance_cost, parking_charges
    ]).reshape(1, -1)
    
    # Check if the model is expecting a different set of features
    if hasattr(best_model, 'feature_names_in_'):
        expected_features = best_model.feature_names_in_
        print(f"Model expects features: {expected_features}")
    
    # Predict the price
    predicted_price = best_model.predict(input_features)
    
    # Generate the feature description
    feature_description = (
        f"Sale year: {sale_year}, Zone: {zone}, Cats allowed: {cats_allowed}, "
        f"Dogs allowed: {dogs_allowed}, Coop/Condo: {coop_condo}, Garage exists: {garage_exists}, "
        f"Bedrooms: {num_bedrooms}, Floors in building: {num_floors_in_building}, "
        f"Full bathrooms: {num_full_bathrooms}, Half bathrooms: {num_half_bathrooms}, "
        f"Total rooms: {num_total_rooms}, Dining room type: {dining_room_type}, Fuel type: {fuel_type}, "
        f"Kitchen type: {kitchen_type}, Community district number: {community_district_num}, "
        f"Approx. year built: {approx_year_built}, Square footage: {sq_footage}, "
        f"Walk score: {walk_score}, Maintenance cost: {maintenance_cost}, Parking charges: {parking_charges}"
    )
    
    return predicted_price[0], feature_description

# Example usage with the provided sample values
sale_year = 2017  # (2016, 2017)
zone = 2  # (1 - 9)
cats_allowed = 1  # (0,1)
dogs_allowed = 1  # (0,1)
coop_condo = 2  # (1:Coop, 2:Condo)
garage_exists = 0  # (0,1)
num_bedrooms = 1  # (1,..,n)
num_floors_in_building = 5  # (1,..,n)
num_full_bathrooms = 1  # (1,..,n)
num_half_bathrooms = 0  # (1,..,n)
num_total_rooms = 1  # (1,...,n)
dining_room_type = 1  # (1:none, 2:dining area, 3:combo, 4:other, 5:formal)
fuel_type = 5  # (1:oil, 2:none, 3:gas, 4:other, 5:electric)
kitchen_type = 0  # (1: none, 2:efficiency, 3:other, 4:eatin, 5:combo)
community_district_num = 26
approx_year_built = 1950  # (19xx - 20xx)
sq_footage = 1200
walk_score = 99
maintenance_cost = 500
parking_charges = 0

# Assuming best_model is already defined and trained
predicted_price, feature_description = predict_price(
    sale_year, zone, cats_allowed, dogs_allowed, coop_condo, garage_exists,
    num_bedrooms, num_floors_in_building, num_full_bathrooms, num_half_bathrooms,
    num_total_rooms, dining_room_type, fuel_type, kitchen_type, community_district_num,
    approx_year_built, sq_footage, walk_score, maintenance_cost, parking_charges
)
formatted_price = f"${predicted_price:,.0f}"
print(f"Predicted Price: {formatted_price}")
print(f"Feature Description: {feature_description}")


Model expects features: ['sale_year' 'zone' 'cats_allowed' 'dogs_allowed' 'coop_condo'
 'garage_exists' 'num_bedrooms' 'num_floors_in_building'
 'num_full_bathrooms' 'num_half_bathrooms' 'num_total_rooms'
 'dining_room_type' 'fuel_type' 'kitchen_type' 'community_district_num'
 'approx_year_built' 'sq_footage' 'walk_score' 'maintenance_cost'
 'parking_charges']
Predicted Price: $481,731
Feature Description: Sale year: 2017, Zone: 2, Cats allowed: 1, Dogs allowed: 1, Coop/Condo: 2, Garage exists: 0, Bedrooms: 1, Floors in building: 5, Full bathrooms: 1, Half bathrooms: 0, Total rooms: 1, Dining room type: 1, Fuel type: 5, Kitchen type: 0, Community district number: 26, Approx. year built: 1950, Square footage: 1200, Walk score: 99, Maintenance cost: 500, Parking charges: 0


c:\Users\lenovo\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:465: UserWarning: X does not have valid feature names, but BaggingRegressor was fitted with feature names
  warnings.warn(


# DEMO 
## INPUT : 
- `Sale year`: 2017, 
- `Zone:` 2  (North Queens), 
- `Cats allowed`: 1 (yes), 
- `Dogs allowed`: 1 (yes), 
- `Coop/Condo`: 2 (condo), 
- `Garage exists`: 0 (no garage), 
- `Bedrooms`: 1 (1 bedroom), 
- `Floors in building`: 5 (5 total floors),-

- `Full bathrooms:` 1, 
- `Half bathrooms:` 0, 
- `Total  rooms`: 1, 
- `Dining room type`: 1 (none), 
- `Fuel type`: 5 (electric), 
- `Kitchen type`: 1 (none), 
- `Community district number`: 26, 
- `Approx. year built`: 1950, 
- `Square footage`: 1200, 
- `Walk score: 99`, 
- `Maintenance cost`: 500, 
- `Parking charges`: 0

## OUTPUT: $481,731